src/Common/src/System/Security/Cryptography/RSACng.cs

```diff
@@ -55,6 +60,30 @@ protected override byte[] HashData(Stream data, HashAlgorithmName hashAlgorithm)
         {
             return CngCommon.HashData(data, hashAlgorithm);
         }
+
+        private void SetKeySize(int newKeySize)
+        {
+            // If the new keysize isn't considered legal then build a temporary KeySizes array which
+            // includes this size as a one-off.
+            if (!newKeySize.IsLegalSize(s_creationKeySizes))
+            {
+                KeySizes[] newKeySizes = new KeySizes[s_creationKeySizes.Length + 1];
+                Array.Copy(s_creationKeySizes, newKeySizes, s_creationKeySizes.Length);
+                newKeySizes[newKeySizes.Length - 1] = new KeySizes(newKeySize, newKeySize, 0);
+
+                // Now KeySize can be set.
+                s_legalKeySizes = newKeySizes;
+
+                KeySize = newKeySize;
+
+                // And now behavior is restored.
+                s_legalKeySizes = s_creationKeySizes;
```

Comment:
Isn't there a race condition here?  We're changing the value of a static that's shared by all instances... what if multiple instances each come along and call SetKeySize at the same time, with different newKeySizes?